app/forms/openid_connect_token_form.rb

```diff
@@ -91,25 +91,34 @@ def pkce_sp
 
   def validate_pkce_or_private_key_jwt
     return if pkce? || private_key_jwt?
-    errors.add :code, t('openid_connect.token.errors.invalid_authentication')
+    errors.add :code,
+               t('openid_connect.token.errors.invalid_authentication'),
+               type: :invalid_authentication
   end
 
   def validate_expired
     if identity&.updated_at && identity.updated_at < session_expiration
-      errors.add :code, t('openid_connect.token.errors.expired_code')
+      errors.add :code, t('openid_connect.token.errors.expired_code'), type: :expired_code
     end
   end
 
   def validate_code
-    errors.add :code, t('openid_connect.token.errors.invalid_code') if identity.blank? ||
-                                                                       !identity.user
+    if identity.blank? || !identity.user
+      errors.add :code,
+                 t(
+                   'openid_connect.token.errors.invalid_code',
+                 ),
```

Comment:
I think we can simplify this?

```suggestion
                 t('openid_connect.token.errors.invalid_code'),
```